notebooks/ch11/gradient_stochastic.md

```diff
@@ -0,0 +1,331 @@
+$$
+
+In this equation:
+- $\hat\theta_{t}$ is our current estimate of $\theta$ at the $t$th iteration
+- $\alpha$ is the learning rate
+- $\nabla_{\hat\theta} L(\hat\theta, y)$ is the gradient of the loss function
+- We compute the next estimate $\hat\theta_{t+1}$ by subtracting the product of $\alpha$ and $\nabla_{\hat\theta} L(\hat\theta, y)$ computed at $\hat\theta_{t}$
+
+## Limitations of Gradient Descent
+
+Often, we calculate $\nabla_{\hat\theta}L(\hat\theta, y)$ using the average gradient of the loss function $l(\hat\theta, y_i)$ across the entire dataset as a single batch. For this reason, this gradient update rule is often referred to as **batch gradient descent**.
+
+For example, the gradient of the MSE loss first requires us to find the gradient of the squared loss, $\nabla_{\hat\theta} l(\hat\theta, y_i) = -2 (y_i - \hat\theta)$, for each of the $n$ observations in our dataset. We then compute the average of these results. This is shown below:
+
+$$
+\begin{align}
+\nabla_{\hat\theta} L(\hat\theta, y) &= -\frac{2}{n} \sum_{i=1}^{n}(y_i - \hat\theta) \\
+&= \frac{1}{n} \sum_{i=1}^{n}-2(y_i - \hat\theta) \\
+&= \frac{1}{n} \sum_{i=1}^{n} \nabla_{\hat\theta} l(\hat\theta, y_i)
+\end{align}
+$$
+
+
+
+This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve.
+
+## Stochastic Gradient Descent
+
+To circumvent the difficulty of computing a gradient across the entire training set, **stochastic gradient descent** approximates the overall gradient using a single randomly chosen data point. Since the observation is chosen randomly, we expect that using the gradient at each individual observation will eventually converge the algorithm to the same parameters as batch gradient descent. The general update formula for stochastic gradient descent is below, where $l(\hat\theta, y_i)$ is the loss function for a single data point:
+
+$$
+\hat\theta_{t+1} = \hat\theta_t - \alpha \nabla_\hat\theta l(\hat\theta, y_i)
+$$
+
+Returning back to our example using the MSE, we approximate the gradient of the mean squared error using the gradient of the squared loss of one data point. 
+
+$$
+\begin{align}
+\nabla_{\hat\theta}L(\hat\theta, y) &\approx \nabla_{\hat\theta} l(\hat\theta, y_i) \\
+&= -2(y_i - \hat\theta)
+\end{align}
+$$
+
+Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded since the randomness means that $E[\nabla_{\hat\theta}l(\hat\theta, y_i)] = \nabla_{\hat\theta}L(\hat\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through all $n$ observations of the shuffled data is called an **epoch**; at the end of every epoch, we re-shuffle the data to ensure that the model sees data points in an arbitrary order.
```

Comment:
We can't include the second sentence since we haven't introduced expectation at this point—let's just omit it.

"arbitrary order" and "random order" have very different meanings in statistics. I think we should just end the last sentence with "we reshuffle the data". (Technically the data are already in a random order after the first shuffle.)